pkg/platform/runtime/buildscript/buildscript.go

```diff
@@ -135,38 +138,35 @@ func (s *Script) String() string {
 	buf.WriteString("\n")
 	buf.WriteString("main = ")
 	switch {
-	case s.Expr.Let.In.FuncCall != nil:
-		buf.WriteString(apString(s.Expr.Let.In.FuncCall))
-	case s.Expr.Let.In.Name != nil:
-		buf.WriteString(*s.Expr.Let.In.Name)
+	case s.expr.Let.In.FuncCall != nil:
+		buf.WriteString(apString(s.expr.Let.In.FuncCall))
+	case s.expr.Let.In.Name != nil:
+		buf.WriteString(*s.expr.Let.In.Name)
 	}
 
 	return buf.String()
 }
 
-// prepareForOutput prepares this script's buildexpression for output in buildscript format.
-// For example, substitute the solve node's "at_time" timestamp with a variable reference.
-// Since this function directly modifies the Expr field, callers of this function must balance it
-// with a call to `prepareForUse()` to restore the Expr to a useful state.
-func (s *Script) prepareForOutput() {
-	if s.atTime != nil {
-		_, err := s.Expr.SetDefaultTimestamp()
-		if err != nil {
-			multilog.Error("Failed to set default timestamp: %v", err)
-		}
-	}
-}
-
-// prepareForUse prepares this script's buildexpression for use in a buildplanner context.
+// BuildExpression returns a copy of this script's underlying buildexpression for use in a
+// buildplanner context.
 // For example, the solve node's "at_time" will not contain a variable reference if this script
 // has a top-level assignment for it.
-func (s *Script) prepareForUse() {
+func (s *Script) BuildExpression() (*buildexpression.BuildExpression, error) {
```

Comment:
An elegant solution to a dirty problem, I like it!

I wonder if we should name this something more explicit though, to communicate that it's not the "raw" build expression? eg. `ProcessedBuildExpression()`.

Reply:
I don't think clients need to be aware that there's a distinction. buildscript's internals care, sure, but outside of buildscripts, I don't think anyone needs to know.